book/src/rules/relations.md

```diff
@@ -0,0 +1,137 @@
+
+Let's start with the simplest relation: the representation of the Control Flow Graph, in the `cfg_edge` relation.
+
+Note: some of these relations are in the process of being renamed. This document will be updated when that happens.
+
+### 1. `cfg_edge`
+
+`cfg_edge(node1, node2)`: as its name suggests, this relation stores that there's a CFG edge between the node `node1` and the node `node2`.
+
+For each MIR statement location, 2 Polonius nodes are generated: the "Start" nodes and the "Mid" nodes (some of the other Polonius inputs will later be recorded at each of the nodes). These 2 nodes are linked by an edge recorded in this relation.
+
+Then, another edge will be recorded, linking this MIR statement to its successor statement(s): from the mid node of the current location to the start node of the successor location. Even though it's encoded differently in MIR, this will similarly apply when the successor location is in another block, linking the mid node of the current location to the start node of the successor block's starting location.
+
+For example, for this MIR (edited from the example for clarity, and to only show the parts related to the CFG):
+
+```rust
+bb0: {
+  ...          // bb0[0]
+  ...          // bb0[1]
+  goto -> bb3; // bb0[2]
+}
+
+...
+
+bb3: {
+  ... // bb3[0]
+}
+
+```
+
+we will record these input facts (as mentioned before, they'll be interned) in the `cfg_edge` relation, shown here as pseudo Rust:
+
+```rust
+cfg_edge = vec![
+  // statement at location bb0[0]:
+  (bb0-0-start, bb0-0-mid),
+  (bb0-0-mid, bb0-1-start),
+  
+  // statement at location bb0[1]:
+  (bb0-1-start, bb0-1-mid),
+  (bb0-1-mid, bb0-2-start),
+  
+  // terminator at location bb0[2]:
+  (bb0-2-start, bb0-2-mid),
+  (bb0-2-mid, bb3-0-start),
+];
+```
+
+### 2. `borrow_region`
+
+`borrow_region(origin, loan, node)`: this relation stores that the origin `origin` may refer to data from loan `loan` from the node `node` and onwards.
+
+For every borrow expression, a loan will be created and there will be a fact stored in this relation to link this loan to the origin of the borrow expression.
+
+For example
+
+```rust
+let mut a = 0;
+let r = &mut a; // this creates the loan L0
+//      ^ let's call this 'a
+```
+
+there will be a `borrow_region` fact linking `L0` to `'a` at this node. This loan will flow along the CFG and the subset relationships between origins, and the computation will require that its terms are respected or it will generate an illegal access error.
+
+### 3. `universal_region`
+
+`universal_region(origin)`: this relation stores that the origin `origin` is a "universal region"/"free region"/"placeholder region" and not defined in the MIR body we're checking.
+
+Those are the default universal origins (`'static`) and the ones defined on functions which are generic over a lifetime. For example, with
```

Comment:
and here as well: should we call these universal origins, universal regions, free regions, free origins, etc ?